pkg/check/gsoc/gsoc.go

```diff
@@ -0,0 +1,347 @@
+
+// Check instance.
+type Check struct {
+	logger logging.Logger
+}
+
+type socData struct {
+	Owner string
+	Sig   string
+	Data  []byte
+}
+
+// NewCheck returns a new check instance.
+func NewCheck(logger logging.Logger) beekeeper.Action {
+	return &Check{
+		logger: logger,
+	}
+}
+
+func (c *Check) Run(ctx context.Context, cluster orchestration.Cluster, opts interface{}) (err error) {
+	o, ok := opts.(Options)
+	if !ok {
+		return fmt.Errorf("invalid options type")
+	}
+
+	fullNodeNames := cluster.FullNodeNames()
+	clients, err := cluster.NodesClients(ctx)
+	if err != nil {
+		return err
+	}
+
+	if len(fullNodeNames) < 2 {
+		return fmt.Errorf("gsoc test require at least 2 full nodes")
+	}
+
+	uploadClient := clients[fullNodeNames[0]]
+	listenClient := clients[fullNodeNames[1]]
+
+	batches := make([]string, 2)
+	for i := 0; i < 2; i++ {
+		c.logger.Infof("gsoc: creating postage batch. amount=%d, depth=%d, label=%s", o.PostageAmount, o.PostageDepth, o.PostageLabel)
+		batchID, err := uploadClient.CreatePostageBatch(ctx, o.PostageAmount, o.PostageDepth, o.PostageLabel, false)
+		if err != nil {
+			return err
+		}
+		c.logger.Infof("gsoc: postage batch created: %s", batchID)
+		batches[i] = batchID
+	}
+
+	c.logger.Infof("send messages with different postage batches sequentially...")
+	err = run(ctx, uploadClient, listenClient, batches, c.logger, false)
+	if err != nil {
+		return err
+	}
+	c.logger.Infof("done")
+
+	c.logger.Infof("send messages with different postage batches parallel...")
+	err = run(ctx, uploadClient, listenClient, batches, c.logger, true)
+	if err != nil {
+		return err
+	}
+	c.logger.Infof("done")
+
+	return nil
+}
+
+func run(ctx context.Context, uploadClient *bee.Client, listenClient *bee.Client, batches []string, logger logging.Logger, parallel bool) error {
+	const numChunks = 10
+	privKey, err := crypto.GenerateSecp256k1Key()
+	if err != nil {
+		return err
+	}
+
+	addresses, err := listenClient.Addresses(ctx)
+	if err != nil {
+		return err
+	}
+	resourceId, socAddress, err := mineResourceId(ctx, addresses.Overlay, privKey, 1)
+	if err != nil {
+		return err
+	}
+	logger.Infof("gsoc: socAddress=%s, listner node address=%s", socAddress, addresses.Overlay)
+
+	listener := &socListener{}
+	ch, err := listener.Listen(ctx, listenClient.Config().APIURL.Host, socAddress, logger)
+	if err != nil {
+		return fmt.Errorf("listen websocket: %w", err)
+	}
+	defer listener.Close()
+
+	received := make(map[string]bool, numChunks)
+	receivedMtx := new(sync.Mutex)
+
+	go func() {
+		for p := range ch {
+			receivedMtx.Lock()
+			received[p] = true
+			receivedMtx.Unlock()
+		}
+	}()
+
+	if parallel {
+		err = runInParallel(ctx, uploadClient, numChunks, batches, resourceId, privKey, logger)
+	} else {
+		err = runInSequence(ctx, uploadClient, numChunks, batches, resourceId, privKey, logger)
+	}
+	if err != nil {
+		return err
+	}
+
+	// wait for listener to receive all messages
+	time.Sleep(5 * time.Second)
```

Comment:
Does this mean that all messages are expected to be received within this timeframe?

Reply:
yes